src/auto-evo/steps/ForceExtinction.cs

```diff
@@ -0,0 +1,80 @@
+﻿namespace AutoEvo
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Godot;
+
+    /// <summary>
+    ///   Forces extinction of worse-faring species to limit the number of steps in next auto-evo run
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This was a hotfix to the 0.5.6 release.
+    ///     TODO: Inquire more refined solution, such as random extinction with bias for smaller populations.
+    ///   </para>
+    /// </remarks>
+    public class ForceExtinction : IRunStep
+    {
+        private readonly List<Patch> patches;
+        private readonly AutoEvoConfiguration configuration;
+
+        public ForceExtinction(List<Patch> patches, AutoEvoConfiguration configuration)
+        {
+            this.patches = patches;
+            this.configuration = configuration;
+        }
+
+        public bool CanRunConcurrently => false;
+
+        public int TotalSteps => 1;
+
+        public bool RunStep(RunResults results)
+        {
+            var populationsByPatch = results.GetPopulationsByPatch(!configuration.ProtectMigrationsFromSpeciesCap,
+                !configuration.AllowNoMutation, !configuration.ProtectNewCellsFromSpeciesCap);
+
+            foreach (Patch patch in patches)
+            {
+                var speciesInPatch = populationsByPatch[patch];
+
+                // This does not take player into account as the species can never go extinct this way.
+                var protectedSpeciesCount = 0;
+
+                if (configuration.ProtectNewCellsFromSpeciesCap)
+                    protectedSpeciesCount += results.GetNewSpeciesResults(patch).Count;
+                if (configuration.ProtectMigrationsFromSpeciesCap)
+                    protectedSpeciesCount += results.GetMigrationsTo(patch).Count;
+
+                // Only bother if we're above the limit
+                if (speciesInPatch.Count + protectedSpeciesCount <= configuration.MaximumSpeciesInPatch)
+                    continue;
+
+                GD.Print("Running extinction step in patch ", patch.Name, ". ",
+                    "Total count:", speciesInPatch.Count + protectedSpeciesCount);
+
+                // Sort the species in the patch, unless protected species fill up all the place already...
+                var orderedSpeciesInPatch = (configuration.MaximumSpeciesInPatch > protectedSpeciesCount) ?
+                    speciesInPatch.OrderBy(s => s.Value).Select(s => s.Key) :
+                    speciesInPatch.Keys;
+
+                var speciesToRemoveCount = speciesInPatch.Count - Math.Max(
+                    configuration.MaximumSpeciesInPatch - protectedSpeciesCount, 0);
+
+                // Remove worst-faring species, except for the player's and protected species
+                foreach (var speciesToRemove in orderedSpeciesInPatch.Take(speciesToRemoveCount))
+                {
+                    if (speciesToRemove.PlayerSpecies)
+                        continue;
```

Comment:
Because you limit things with Take, I think continue here, makes the counts inaccure. Instead I'd just loop without the .Take and have a separate variable tracking how many things are left to remove, and the loop is broken early once that reaches 0.

Reply:
No, this is intended: you don't penalize other species if the player is not doing good. So if you keep only first 10 species and the player is 11th, you still keep the first 10. You just also keep the player.

Reply:
Ah so you already computed speciesToRemoveCount to take the player into account?
Reading things again, I'm not seeing that. You would need to somehow skip the protected species before you .Take, otherwise this logic just does not sit right with me.

Reply:
No. There are two cases (assuming the threshold is 10)

a) the player is in the top 10 species with most pop, in which case everything's normal, since they are not in the list to remove.
b) the player isn't and as such they are into the list to remove. Then you end up with 11 species remaining : the top 10 + the player. 
This is an exception to the 10 species limit, since it only concerns 1 single species that is consistently spared (unlike a new species that is at best only spared once).

Reply:
But what about when there are more than 1 protected species due to the moves etc.? Wouldn't then end up not doing the right thing?

Reply:
Other protected species, which are protected only for one generation, are excluded of `orderedSpeciesInPatch`, and are handled by another logic (I should change the var name, actually)

In the end, unless you only have protected species, you have, at most `cap` + 1 (the player's) species.

Reply:
Okay, maybe I can now understand this. Though, that variable renaming and perhaps one or two more comments would be welcome to make this clear to future me and other future developers.

Reply:
I did change a few variable names and added comments. Do you think it is clear enough now?